codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -857,9 +876,15 @@ public Http2Stream closeRemoteSide() {
         }
 
         @Override
-        public int nextStreamId() {
-            // For manually created client-side streams, 1 is reserved for HTTP upgrade, so start at 3.
-            return nextStreamId > 1 ? nextStreamId : nextStreamId + 2;
+        public int incrementAndGetNextStreamId() {
+            return nextReservationStreamId >= 0 ? nextReservationStreamId += 2 : nextReservationStreamId;
+        }
+
+        private void incrementExpectedStreamId(int streamId) {
+            if (streamId > nextReservationStreamId && nextReservationStreamId >= 0) {
```

Comment:
Perhaps this would be more readable if we brought back `isExhausted` even if it's private:

```java
if (!isExhausted() && streamId > nextReservationStreamId()) {
    ...
}
```

... seems a little more clear.  WDYT?